From fc6ff194ee5d2ad1379193fabc485f47bee29b25 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Wed, 14 Nov 2018 16:51:40 +0200
Subject: [PATCH 0792/1051] net: mvpp2: flow control tx stop queue only on
 matched cpu

Apply flow control TXQ-stop (pause) only if TXQ-number match to
(cpu)thread-queue-number (e.g. both are in XPS mapping).

The un-match is probable since some packets may be sent on mapping
different from configured in XPS.
Next "matching" packet would go over stop or in worst case dropped.

Change-Id: I1abe4774fd0335f1072d4f30460ecfa77f0001b6
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/60916
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
index e47619a22f12..b68edeb8fc06 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_main.c
@@ -3821,7 +3821,9 @@ static netdev_tx_t mvpp2_tx(struct sk_buff *skb, struct net_device *dev)
 			mvpp2_aggr_txq_pend_desc_add(port, frags);
 		}
 
-		if (unlikely(txq_pcpu->count >= txq_pcpu->stop_threshold)) {
+		if (unlikely(txq_pcpu->count >= txq_pcpu->stop_threshold) &&
+		    thread == txq_id) {
+			/* Don't stop TXQ not related to this cpu/thread */
 			nq = netdev_get_tx_queue(dev, txq_id);
 			netif_tx_stop_queue(nq);
 		}
-- 
2.17.1

